pymc3/diagnostics.py

```diff
@@ -136,43 +137,40 @@ def gelman_rubin(mtrace):
 
     if mtrace.nchains < 2:
         raise ValueError(
-            'Gelman-Rubin diagnostic requires multiple chains of the same length.')
+            'Gelman-Rubin diagnostic requires multiple chains '
+            'of the same length.')
 
-    def calc_rhat(x):
-
-        try:
-            # When the variable is multidimensional, this assignment will fail, triggering
-            # a ValueError that will handle the multidimensional case
-            m, n = x.shape
-
-            # Calculate between-chain variance
-            B = n * np.var(np.mean(x, axis=1), ddof=1)
+    Rhat = {}
+    for var in mtrace.varnames:
+        x = np.array(mtrace.get_values(var, combine=False))
 
-            # Calculate within-chain variance
-            W = np.mean(np.var(x, axis=1, ddof=1))
+        # make sure to handle scalars correctly - add extra dim if needed
+        if len(x.shape) == 2:
+            is_scalar = True
+            x = np.atleast_3d(mtrace.get_values(var, combine=False))
+        else:
+            is_scalar = False
```

Comment:
This entire block could be eliminate if we were okay potentially returning array(0.6) like objects. I am not sure off the top of my head if this would happen.